mitiq/calibration/calibrator.py

```diff
@@ -178,12 +180,49 @@ def best_strategy(self, results: List[Dict[str, Any]]) -> Strategy:
             parameters that performed best.
         """
         best_improvement_factor = 1.0
+        num_circuits = len(self.settings.circuit_types)
+
         strategy = DefaultStrategy
-        for circuit in results:
-            for result in circuit["mitigated_values"]["ZNE"]["results"]:
-                if result["improvement_factor"] > best_improvement_factor:
-                    best_improvement_factor = result["improvement_factor"]
-                    strategy = result["strategy"]
+
+        def filter_on_strategy(
+            result: Dict[str, Dict[str, Dict[str, Any]]], strategy_id: int
+        ) -> Dict[str, Dict[str, Dict[str, Any]]]:
+            res = result
+            res["mitigated_values"]["ZNE"]["results"] = [
+                res["mitigated_values"]["ZNE"]["results"][strategy_id]
+            ]
+            return result
+
+        for strategy_id in range(
+            len(results[0]["mitigated_values"]["ZNE"]["results"])
+        ):
+            strategy_group = []
+            for c in range(num_circuits):
+                result_copy = deepcopy(results)
+                strategy_group.append(
+                    filter_on_strategy(result_copy[c], strategy_id=strategy_id)
+                )
+                self.compute_improvements(strategy_group)
+
+            if (
+                strategy_group[0]["mitigated_values"]["ZNE"]["results"][0][
+                    "improvement_factor"
+                ]
+                > best_improvement_factor
+            ):
+                best_improvement_factor = strategy_group[0][
+                    "mitigated_values"
+                ]["ZNE"]["results"][0]["improvement_factor"]
+                strategy = strategy_group[0]["mitigated_values"]["ZNE"][
+                    "results"
+                ][0]["strategy"]
+        self.results.append(
+            {"best_improvement_factor": best_improvement_factor}
+        )
```

Comment:
It seems like this is working (!), but there are a few things I'm hesitant about here.

1. We shouldn't rely on the ordering of the strategies to be the same in each `["mitigated_values"]["ZNE"]["results"]` chain.
2. `compute_improvements` was written to be called on the output of `run_circuits`. Does `strategy_group` have the same structure?
3. There are some really long chains going into this dictionary here, which I think is impossible to avoid, but I think setting some as variables will be helpful for readability and maintainability. E.g. on line 197, that variable we are using `range` on could be called `num_strategies`.
4. I think this loop could be simplified quite a bit by using a dictionary to keep track of things.
```py
strategy_improvement_factors: Dict[Strategy, List[float]] = defaultdict(list) # not sure what the type would be exactly
for circuit in results:
    for result in circuit["mitigated_values"]["ZNE"]["results"]:
        strategy = result["strategy"]
        value = result["mitigated_value"] # or could be improvement factor
        strategy_improvement_factors[strategy].append(value)
```
Then once you have `strategy_improvement_factors` you can write a function to compute the average IF, or do it as in `compute_improvements` if you store the ideal and noisy values in the array as well. What do you think of this approach?